feat(cognito-idp): Code Generation for Cognito IdentityProvider

Benedikt1992 opened the pull request:
<!--
Thank you for helping to improve Crossplane!

Please read through https://git.io/fj2m9 if this is your first time opening a
Crossplane pull request. Find us in https://slack.crossplane.io/messages/dev if
you need any help contributing.
-->

### Description of your changes

<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Crossplane issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

-->
Fixes #720

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested
This code has been tested by the unit tests and manually with the included examples

<!--
Before reviewers can be confident in the correctness of this pull request, it
needs to tested and shown to be correct. Briefly describe the testing that has
already been done or which is planned for this change.
-->

[contribution process]: https://git.io/fj2m9

haarchri commented on apis/cognitoidentityprovider/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,124 @@
+	// +immutable
+	// +crossplane:generate:reference:type=github.com/crossplane/provider-aws/apis/iam/v1beta1.Role
+	// +crossplane:generate:reference:extractor=github.com/crossplane/provider-aws/apis/iam/v1beta1.RoleARN()
+	RoleARN *string `json:"roleARN,omitempty"`
```
can we use here `RoleARN *string `json:"roleArn,omitempty"``

haarchri commented on apis/cognitoidentityprovider/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,124 @@
+
+	// UserPoolIDRef is a reference to an server instance.
+	// +optional
+	RoleARNRef *xpv1.Reference `json:"roleARNRef,omitempty"`
```
RoleARNRef *xpv1.Reference `json:"roleArnRef,omitempty"`

haarchri commented on apis/cognitoidentityprovider/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,124 @@
+
+	// UserPoolIDSelector selects references to an server instance.
+	// +optional
+	RoleARNSelector *xpv1.Selector `json:"roleARNSelector,omitempty"`
```
RoleARNSelector *xpv1.Selector `json:"roleArnSelector,omitempty"`

haarchri commented on apis/cognitoidentityprovider/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,124 @@
+	// The user pool ID.
+	// +immutable
+	// +crossplane:generate:reference:type=UserPool
+	UserPoolID *string `json:"userPoolID,omitempty"`
```
UserPoolID *string `json:"userPoolId,omitempty"`

haarchri commented on apis/cognitoidentityprovider/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,124 @@
+
+	// UserPoolIDRef is a reference to an server instance.
+	// +optional
+	UserPoolIDRef *xpv1.Reference `json:"userPoolIDRef,omitempty"`
```
UserPoolIDRef *xpv1.Reference `json:"userPoolIdRef,omitempty"`

haarchri commented on apis/cognitoidentityprovider/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,124 @@
+
+	// UserPoolIDSelector selects references to an server instance.
+	// +optional
+	UserPoolIDSelector *xpv1.Selector `json:"userPoolIDSelector,omitempty"`
```
UserPoolIDSelector *xpv1.Selector `json:"userPoolIdSelector,omitempty"`

haarchri commented on apis/cognitoidentityprovider/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,124 @@
+	// The user pool ID.
+	// +immutable
+	// +crossplane:generate:reference:type=UserPool
+	UserPoolID *string `json:"userPoolID,omitempty"`
```
UserPoolID *string `json:"userPoolId,omitempty"`

haarchri commented on apis/cognitoidentityprovider/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,124 @@
+
+	// UserPoolIDRef is a reference to an server instance.
+	// +optional
+	UserPoolIDRef *xpv1.Reference `json:"userPoolIDRef,omitempty"`
```
UserPoolIDRef *xpv1.Reference `json:"userPoolIdRef,omitempty"`

haarchri commented on apis/cognitoidentityprovider/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,124 @@
+
+	// UserPoolIDSelector selects references to an server instance.
+	// +optional
+	UserPoolIDSelector *xpv1.Selector `json:"userPoolIDSelector,omitempty"`
```
UserPoolIDSelector *xpv1.Selector `json:"userPoolIdSelector,omitempty"`

haarchri commented on apis/cognitoidentityprovider/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,124 @@
+	// The user pool ID.
+	// +immutable
+	// +crossplane:generate:reference:type=UserPool
+	UserPoolID *string `json:"userPoolID,omitempty"`
```
UserPoolID *string `json:"userPoolId,omitempty"`

haarchri commented on apis/cognitoidentityprovider/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,124 @@
+
+	// UserPoolIDRef is a reference to an server instance.
+	// +optional
+	UserPoolIDRef *xpv1.Reference `json:"userPoolIDRef,omitempty"`
```
UserPoolIDRef *xpv1.Reference `json:"userPoolIdRef,omitempty"`

haarchri commented on apis/cognitoidentityprovider/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,124 @@
+
+	// UserPoolIDSelector selects references to an server instance.
+	// +optional
+	UserPoolIDSelector *xpv1.Selector `json:"userPoolIDSelector,omitempty"`
```
UserPoolIDSelector *xpv1.Selector `json:"userPoolIdSelector,omitempty"`

haarchri commented on apis/cognitoidentityprovider/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,124 @@
+	// The user pool ID.
+	// +immutable
+	// +crossplane:generate:reference:type=UserPool
+	UserPoolID *string `json:"userPoolID,omitempty"`
```
UserPoolID *string `json:"userPoolId,omitempty"`

haarchri commented on apis/cognitoidentityprovider/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,124 @@
+
+	// UserPoolIDRef is a reference to an server instance.
+	// +optional
+	UserPoolIDRef *xpv1.Reference `json:"userPoolIDRef,omitempty"`
```
UserPoolIDRef *xpv1.Reference `json:"userPoolIdRef,omitempty"`

haarchri commented on apis/cognitoidentityprovider/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,124 @@
+
+	// UserPoolIDSelector selects references to an server instance.
+	// +optional
+	UserPoolIDSelector *xpv1.Selector `json:"userPoolIDSelector,omitempty"`
```
UserPoolIDSelector *xpv1.Selector `json:"userPoolIdSelector,omitempty"`

haarchri commented on examples/cognito/group.yaml:
```diff
@@ -0,0 +1,14 @@
+  name: example-group
+spec:
+  forProvider:
+    region: eu-central-1
```
`region: us-east-1`

haarchri commented on examples/cognito/identityprovider.yaml:
```diff
@@ -0,0 +1,52 @@
+  name: example-provider
+spec:
+  forProvider:
+    region: eu-central-1
```
`region: us-east-1`

haarchri commented on examples/cognito/identityprovider.yaml:
```diff
@@ -0,0 +1,52 @@
+    # You'll need to update this to details specified by your external identity provider
+    providerDetails:
+      client_id: foo000000000000000000
+      client_secret: bar000000000000000000
```
think providerDetails should be a kubernetes secret ?! like your other example - i would skip normal providerDetails because of secret content ?

haarchri commented on examples/cognito/identityprovider.yaml:
```diff
@@ -0,0 +1,52 @@
+  name: example-provider-with-secret
+spec:
+  forProvider:
+    region: eu-central-1
```
`region: us-east-1`

haarchri commented on examples/cognito/userpoolclient.yaml:
```diff
@@ -0,0 +1,15 @@
+  name: example-client
+spec:
+  forProvider:
+    region: eu-central-1
```
`region: us-east-1`

haarchri commented on examples/cognito/userpooldomain.yaml:
```diff
@@ -0,0 +1,11 @@
+  name: example-domain
+spec:
+  forProvider:
+    region: eu-central-1
```
`region: us-east-1`

haarchri commented on pkg/controller/cognitoidentityprovider/identityprovider/setup.go:
```diff
@@ -0,0 +1,163 @@
+	}
+	switch {
+	case !reflect.DeepEqual(cr.Spec.ForProvider.AttributeMapping, provider.AttributeMapping),
+		!reflect.DeepEqual(cr.Spec.ForProvider.IDpIdentifiers, provider.IdpIdentifiers),
```
`cr.Spec.ForProvider.IDpIdentifiers ` think this is related to ACK - that we have here `IDp` insted of `Idp` ?

Benedikt1992 replied:
Yes, this is due to ACK

haarchri reviewed (changes requested):
added a few small comments

haarchri commented:
i tested the implementation and found 2 issues

```
NAME                                                            READY   SYNCED   EXTERNAL-NAME
group.cognitoidentityprovider.aws.crossplane.io/example-group   True    True     example-group

NAME                                                         READY   SYNCED   EXTERNAL-NAME
userpool.cognitoidentityprovider.aws.crossplane.io/example   True    True     us-east-1_7bNN4dOUJ

NAME                                                                                      READY   SYNCED   EXTERNAL-NAME
identityprovider.cognitoidentityprovider.aws.crossplane.io/example-provider-with-secret   False   True     example-provider-with-secret

NAME                                                                      READY   SYNCED   EXTERNAL-NAME
userpoolclient.cognitoidentityprovider.aws.crossplane.io/example-client   True    True     3anj8d5ujfd3vqsuo3nefffbh4

NAME                                                                    READY   SYNCED   EXTERNAL-NAME
userpooldomain.cognitoidentityprovider.aws.crossplane.io/example-xetu   True    True     example-xetu
```

found two issues: 
- identityprovider never reached ready true status but is available in amazon
<img width="1021" alt="image" src="https://user-images.githubusercontent.com/21083497/157404065-5f4c8caf-4ffd-489f-9ef7-535bb2a65718.png">

- identityprovider atStatus object is with sensitive informations:

```
Status:
  At Provider:
    Creation Date:       2022-03-09T08:34:39Z
    Last Modified Date:  2022-03-09T08:34:39Z
    Provider Details:
      attributes_request_method:      POST
      attributes_url_add_attributes:  false
      authorize_scopes:               openid profile
      client_id:                      yrmdly79y2O_ovEoLHIAxyxC
      client_secret:                  eFAE90QgeN3n0IxElYW-7gWZzAyjj8WQe3Ul9XrWLDRqbATT
      oidc_issuer:                    https://authorization-server.com
    Provider Name:                    example-provider-with-secret
    User Pool ID:                     us-east-1_7bNN4dOUJ
  Conditions:
    Last Transition Time:  2022-03-09T08:34:39Z
    Reason:                Creating
    Status:                False
    Type:                  Ready
    Last Transition Time:  2022-03-09T08:34:39Z
    Reason:                ReconcileSuccess
    Status:                True
    Type:                  Synced
Events:
  Type     Reason                       Age   From                                                                Message
  ----     ------                       ----  ----                                                                -------
  Normal   CreatedExternalResource      83s   managed/identityprovider.cognitoidentityprovider.aws.crossplane.io  Successfully requested creation of external resource
  Warning  CannotUpdateManagedResource  83s   managed/identityprovider.cognitoidentityprovider.aws.crossplane.io  Operation cannot be fulfilled on identityproviders.cognitoidentityprovider.aws.crossplane.io "example-provider-with-secret": the object has been modified; please apply your changes to the latest version and try again
```

Benedikt1992 commented:
> i tested the implementation and found 2 issues
> 
> ```
> NAME                                                            READY   SYNCED   EXTERNAL-NAME
> group.cognitoidentityprovider.aws.crossplane.io/example-group   True    True     example-group
> 
> NAME                                                         READY   SYNCED   EXTERNAL-NAME
> userpool.cognitoidentityprovider.aws.crossplane.io/example   True    True     us-east-1_7bNN4dOUJ
> 
> NAME                                                                                      READY   SYNCED   EXTERNAL-NAME
> identityprovider.cognitoidentityprovider.aws.crossplane.io/example-provider-with-secret   False   True     example-provider-with-secret
> 
> NAME                                                                      READY   SYNCED   EXTERNAL-NAME
> userpoolclient.cognitoidentityprovider.aws.crossplane.io/example-client   True    True     3anj8d5ujfd3vqsuo3nefffbh4
> 
> NAME                                                                    READY   SYNCED   EXTERNAL-NAME
> userpooldomain.cognitoidentityprovider.aws.crossplane.io/example-xetu   True    True     example-xetu
> ```
> 
> found two issues:
> 
> * identityprovider never reached ready true status but is available in amazon
> 
> <img alt="image" width="1021" src="https://user-images.githubusercontent.com/21083497/157404065-5f4c8caf-4ffd-489f-9ef7-535bb2a65718.png">
> 
> * identityprovider atStatus object is with sensitive informations:
> 
> ```
> Status:
>   At Provider:
>     Creation Date:       2022-03-09T08:34:39Z
>     Last Modified Date:  2022-03-09T08:34:39Z
>     Provider Details:
>       attributes_request_method:      POST
>       attributes_url_add_attributes:  false
>       authorize_scopes:               openid profile
>       client_id:                      yrmdly79y2O_ovEoLHIAxyxC
>       client_secret:                  eFAE90QgeN3n0IxElYW-7gWZzAyjj8WQe3Ul9XrWLDRqbATT
>       oidc_issuer:                    https://authorization-server.com
>     Provider Name:                    example-provider-with-secret
>     User Pool ID:                     us-east-1_7bNN4dOUJ
>   Conditions:
>     Last Transition Time:  2022-03-09T08:34:39Z
>     Reason:                Creating
>     Status:                False
>     Type:                  Ready
>     Last Transition Time:  2022-03-09T08:34:39Z
>     Reason:                ReconcileSuccess
>     Status:                True
>     Type:                  Synced
> Events:
>   Type     Reason                       Age   From                                                                Message
>   ----     ------                       ----  ----                                                                -------
>   Normal   CreatedExternalResource      83s   managed/identityprovider.cognitoidentityprovider.aws.crossplane.io  Successfully requested creation of external resource
>   Warning  CannotUpdateManagedResource  83s   managed/identityprovider.cognitoidentityprovider.aws.crossplane.io  Operation cannot be fulfilled on identityproviders.cognitoidentityprovider.aws.crossplane.io "example-provider-with-secret": the object has been modified; please apply your changes to the latest version and try again
> ```

I have removed the providerDetails from the Status field and fixed the Ready state. The lateInit of the IdentityProvider wasn't stored permanently

haarchri reviewed (approved):
tested again - LGTM @Benedikt1992  thanks for implementation & all changes during review 


```
NAME                                                                      READY   SYNCED   EXTERNAL-NAME
userpoolclient.cognitoidentityprovider.aws.crossplane.io/example-client   True    True     3anj8d5ujfd3vqsuo3nefffbh4

NAME                                                                    READY   SYNCED   EXTERNAL-NAME
userpooldomain.cognitoidentityprovider.aws.crossplane.io/example-xetu   True    True     example-xetu

NAME                                                            READY   SYNCED   EXTERNAL-NAME
group.cognitoidentityprovider.aws.crossplane.io/example-group   True    True     example-group

NAME                                                         READY   SYNCED   EXTERNAL-NAME
userpool.cognitoidentityprovider.aws.crossplane.io/example   True    True     us-east-1_7bNN4dOUJ

NAME                                                                                      READY   SYNCED   EXTERNAL-NAME
identityprovider.cognitoidentityprovider.aws.crossplane.io/example-provider-with-secret   True    True     example-provider-with-secret
```
